sdks/java/core/src/main/java/org/apache/beam/sdk/util/ShardedKey.java

```diff
@@ -0,0 +1,117 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.sdk.util;
+
+import com.google.auto.value.AutoValue;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.util.Collections;
+import java.util.List;
+import javax.annotation.Nullable;
+import org.apache.beam.sdk.coders.BooleanCoder;
+import org.apache.beam.sdk.coders.ByteArrayCoder;
+import org.apache.beam.sdk.coders.StructuredCoder;
+import org.apache.beam.sdk.util.common.ElementByteSizeObserver;
+
+/**
+ * A sharded key consisting of a user key and a shard id represented by bytes.
+ *
+ * <p>This is a more generic definition of {@link org.apache.beam.sdk.values.ShardedKey}.
+ */
+@AutoValue
+public abstract class ShardedKey<K> {
+
+  public static <K> ShardedKey<K> of(K key, @Nullable byte[] shardId) {
+    return new AutoValue_ShardedKey(shardId, key);
+  }
+
+  @SuppressWarnings("mutable")
+  @Nullable
+  public abstract byte[] getShardId();
+
+  public abstract K getKey();
+
+  public static class Coder<K> extends StructuredCoder<ShardedKey<K>> {
+
+    private final ByteArrayCoder shardCoder = ByteArrayCoder.of();
+    private final org.apache.beam.sdk.coders.Coder<K> keyCoder;
+
+    private Coder(org.apache.beam.sdk.coders.Coder<K> coder) {
+      keyCoder = coder;
+    }
+
+    public static <K> ShardedKey.Coder<K> of(org.apache.beam.sdk.coders.Coder<K> keyCoder) {
+      return new ShardedKey.Coder<K>(keyCoder);
+    }
+
+    public org.apache.beam.sdk.coders.Coder<K> getKeyCoder() {
+      return keyCoder;
+    }
+
+    @Override
+    public void encode(ShardedKey<K> shardedKey, OutputStream outStream) throws IOException {
+      // The encoding should follow the order:
+      //   null indicator
+      //   length of shard id
+      //   shard id
+      //   encoded user key
+      BooleanCoder.of().encode(shardedKey.getShardId() != null, outStream);
+      if (shardedKey.getShardId() != null) {
+        shardCoder.encode(shardedKey.getShardId(), outStream);
```

Comment:
Currently the shard id is encoded using a ByteArrayCoder which uses a VarInt coder to encode/decode the length in 32bits ([link](https://github.com/apache/beam/blob/master/sdks/java/core/src/main/java/org/apache/beam/sdk/coders/ByteArrayCoder.java#L103)), which is consistent with the parsing logic in the unified worker.

Reply:
The coder specification is underspecified as to how large the var int can be:
https://github.com/apache/beam/blob/eb75286af4c308f4fd85393b4a2bf98a98d6d73e/model/pipeline/src/main/proto/beam_runner_api.proto#L673

Java can't support `byte[]` greater then `2^31-1` in size. Also note that a varint64 coder and varint32 coder share the same encoding for all values from `[0, 2^31-1]` so this is unlikely to become an issue for shard id encoding so can you can ignore this for now.

Reply:
The two encodings are the same for all values in the range we'll be dealing with (and neither Java nor GRPC can handle >4G sized byte arrays) so we should be fine here.

Reply:
Cool! Thanks for confirming.